lib/ui/painting/path.cc

```diff
@@ -67,43 +67,71 @@ void CanvasPath::RegisterNatives(tonic::DartLibraryNatives* natives) {
                      FOR_EACH_BINDING(DART_REGISTER_NATIVE)});
 }
 
-CanvasPath::CanvasPath() {}
+CanvasPath::CanvasPath()
+    : path_tracker_(UIDartState::Current()->GetVolatilePathTracker()),
+      tracked_path_(std::make_shared<VolatilePathTracker::Path>()) {
+  FML_DCHECK(path_tracker_);
+  resetVolatility();
+}
+
+CanvasPath::~CanvasPath() = default;
 
-CanvasPath::~CanvasPath() {}
+void CanvasPath::resetVolatility() {
+  if (!tracked_path_->tracking_volatility) {
+    mutable_path().setIsVolatile(true);
+    tracked_path_->frame_count = 0;
+    tracked_path_->tracking_volatility = true;
+    path_tracker_->Insert(tracked_path_);
+  }
+}
+
+void CanvasPath::ReleaseDartWrappableReference() const {
+  FML_DCHECK(path_tracker_);
+  if (tracked_path_->tracking_volatility) {
```

Comment:
This can happen on any thread right? For example, the thread on which the GC happens. In that case, isn't there a data race between this and the the setter in `resetVolatility`? Maybe you are depending on the fact that once the Dart VM releases the reference to the path, we won't be in a situation where we will call reset volatility. But I am not sure that we can guarantee that case since we could still be holding onto the shared pointer on the native side.

Reply:
Just tracking volatility from the get go in the constructor and not needing the flag (and its unsafe check) should make it so that you don't need the check.

Reply:
I believe this would be safe if the `tracking_volatility` flag was made an `atomic<bool>`.

This relies on the contract that `tracking_volatility` will only be set to true by `resetVolatility`, which only happens during Dart->native method calls that run on the UI thread.  When `ReleaseDartWrappableReference` is called, no such method calls can be active because the Dart object is being GCed.

There is a race against `VolatilePathTracker::OnFrame` which could be running concurrently on the UI thread and might clear `tracking_volatility` while removing the path from the tracker.  However, the worst case is that `ReleaseDartWrappableReference` does an unnecessary`Erase` of a path that was already removed.  If that happens `Erase`/`Drain` will just call `unordered_set::erase` for that path which will do nothing.

However, this is confusing and makes it easy to introduce errors.

The advantage of the check is avoiding the performance cost of the `Erase` (which might take a lock) and the later `Drain` if the GCed path is not tracked.

If the cost is small, another alternative would be to always call `Erase` in `ReleaseDartWrappableReference` and avoid reading `tracking_volatility` on the GC thread.  `Drain` on the UI thread could then safely check `tracking_volatility` in order to avoid doing the `unordered_set::erase` if the path is not tracked.

Reply:
> no such method calls can be active because the Dart object is being GCed.

I was thinking of calls to `CanvasPath::Create` from the native side. But I suppose that doesn't happen today. I may be overthinking cases where these objects are created from the native side.

> If the cost is small, another alternative would be to always call Erase in ReleaseDartWrappableReference and avoid reading tracking_volatility on the GC thread.

IMO, that would be a lot clearer.

Reply:
SG, done. We can always come back to this later with a benchmark if it makes a difference.